src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java

```diff
@@ -68,159 +87,189 @@ public DiskSpoolDAO(NucleusPaths paths) throws IOException {
         url = String.format(DATABASE_CONNECTION_URL, path);
     }
 
+    /**
+     * Initialize the database connection.
+     *
+     * @throws SQLException if db is unable to be created
+     */
+    public void initialize() throws SQLException {
+        try (LockScope ls = LockScope.lock(connectionLock.writeLock())) {
+            close();
+            logger.atDebug().kv("url", url).log("Creating DB connection");
+            connection = getDbInstance();
+        }
+    }
+
+    /**
+     * Close DAO resources.
+     */
+    public void close() {
+        try (LockScope ls = LockScope.lock(connectionLock.readLock())) {
+            if (connection != null) {
+                try {
+                    connection.close();
+                } catch (SQLException e) {
+                    logger.atWarn().kv("url", url).log("Unable to close pre-existing connection");
+                }
+            }
+        }
+    }
+
     /**
      * This method will query the existing database for the existing queue of MQTT request Ids
      * and return them in order.
+     *
      * @return ordered list of the existing ids in the persistent queue
-     * @throws SQLException when fails to get SpoolMessage IDs
+     * @throws SQLException         when fails to get SpoolMessage IDs
+     * @throws InterruptedException if interrupted during execution
      */
-    @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    public Iterable<Long> getAllSpoolMessageIds() throws SQLException {
-        List<Long> currentIds;
-        String query = "SELECT message_id FROM spooler;";
-        try (Connection conn = getDbInstance();
-             PreparedStatement pstmt = conn.prepareStatement(query);
-             ResultSet rs = RetryUtils.runWithRetry(sqlStatementRetryConfig, pstmt::executeQuery,
-                     "get-all-spool-message-ids", logger)) {
-            currentIds = getIdsFromRs(rs);
-        } catch (SQLException e) {
-            checkAndHandleCorruption(e);
-            throw e;
-        } catch (Exception e) {
-            throw new SQLException("Failed to get Spool Message IDs", e);
-        }
-        return currentIds;
+    public Iterable<Long> getAllSpoolMessageIds() throws SQLException, InterruptedException {
+        // TODO don't recreate prepared statements every time
```

Comment:
yes! this is a big optimization opportunity as well

Reply:
:)